engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,660 @@
+package stop
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+	psEvents "github.com/onflow/flow-go/state/protocol/events"
+	"github.com/onflow/flow-go/storage"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of stopping blocks execution.
+// It is intended to work tightly with the Engine, not as a general mechanism or interface.
+//
+// StopControl can stop execution or crash the node at a specific block height. The stop
+// height can be set manually or by the version beacon service event. This leads to some
+// edge cases that are handled by the StopControl:
+//
+//  1. stop is already set manually and is set again manually.
+//     This is considered as an attempt to move the stop height. The resulting stop
+//     height is the new one. Note, the new height can be either lower or higher than
+//     previous value.
+//  2. stop is already set manually and is set by the version beacon.
+//     The resulting stop height is the lower one.
+//  3. stop is already set by the version beacon and is set manually.
+//     The resulting stop height is the lower one.
+//  4. stop is already set by the version beacon and is set by the version beacon.
+//     This means version boundaries were edited. The resulting stop
+//     height is the new one.
+type StopControl struct {
```

Comment:
I think conceptually `StopControl` can be simplified by removing the component part. It seems that `StopControl` is used with Ingestion engine. It's a perfect case for separating engine(as driver, scheduler, orchestrator) and core(actual business logic). We use this approach in other areas and it works very well. Please refer to our recent follower upgrade for [details](https://github.com/onflow/flow-go/blob/master/engine/common/follower/compliance_engine.go). 

The main idea is that `StopControl` implements an interface which is exposed to ingestion engine and it is used purely by it. Ingestion engine implements component manager and provides worker threads for processing blocking events(such as finalization). This will provide a clean interface for `StopControl`, allows unit testing of `StopControl` itself and proper testing of ingestion engine by using mocked object. It also avoids workarounds like [this](https://github.com/onflow/flow-go/pull/4342/files#diff-40f6cbc4e64a72fa73843b6da82561aa729c3c08960f8d97ec6a76bf7d91cd8cR226).

Reply:
I was advised to use a Component. I can look into this, but this will take time. Is it ok if I add this as tech debt?